crates/core/tedge_api/src/plugin.rs

```diff
@@ -0,0 +1,619 @@
+    /// //              t: 15512189350087767644,
+    /// //          },
+    /// //      ),
+    /// //  ],
+    /// // )
+    /// ```
+    ///
+    /// Should you ask for messages the plugin does _not_ support, you will receive a compile
+    /// error:
+    /// ```compile_fail
+    /// # use async_trait::async_trait;
+    /// # use tedge_api::plugin::{Message, Handle, HandleTypes};
+    /// # use tedge_api::PluginError;
+    ///
+    /// struct Heartbeat;
+    ///
+    /// impl Message for Heartbeat {}
+    ///
+    /// struct HeartbeatPlugin;
+    ///
+    /// // This will fail to compile as the `Heartbeat` message is not handled by the plugin
+    /// println!("{:#?}", HandleTypes::get_handlers_for::<(Heartbeat,), HeartbeatPlugin>());
+    /// ```
+    ///
+    /// The error from the rust compiler would look like this (giving a clear indication of what is
+    /// missing):
+    /// ```text
+    /// error[E0277]: the trait bound `HeartbeatPlugin: Handle<Heartbeat>` is not satisfied
+    ///    --> src/plugin.rs:XX:YYY
+    ///     |
+    /// XX  | println!("{:#?}", HandleTypes::get_handlers_for::<(Heartbeat,), HeartbeatPlugin>());
+    ///     |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Handle<Heartbeat>` is not implemented for `HeartbeatPlugin`
+    /// ```
+    pub fn declare_handlers_for<M: MessageBundle, Plugin: DoesHandle<M>>() -> HandleTypes {
+        HandleTypes(M::get_ids())
+    }
+
+    /// Empty list of types. A plugin that does not handle anything will not be able to receive
+    /// messages except through replies sent with [`Reply`](crate::address::Reply)
+    pub fn empty() -> HandleTypes {
+        HandleTypes(Vec::with_capacity(0))
+    }
+}
+
+impl From<HandleTypes> for HashSet<(&'static str, TypeId)> {
+    fn from(ht: HandleTypes) -> Self {
+        ht.0.into_iter().collect()
+    }
+}
+
+/// A thing that can be send around
+///
+/// This trait is a marker trait for all types that can be used as messages which can be send
+/// between plugins in thin-edge.
+pub trait Message: 'static + Send + std::fmt::Debug {
+    type Reply: Message;
+}
+
+/// A bundle of messages
+///
+/// This trait is implemented on types that represent a bundle of different types of messages.
+pub trait MessageBundle {
+    /// Get the names and ids of the types that are represented by this bundle
+    fn get_ids() -> Vec<(&'static str, TypeId)>;
+}
+
+/// An extension for a Plugin implementing type
+///
+/// This trait implements an extension for all types that implement `Plugin`.
+/// This extension can be used by plugin authors to make their specific plugin type instance into a
+/// [`BuiltPlugin`].
+pub trait PluginExt: Plugin {
+    /// Convert a `Plugin` into a `BuiltPlugin`
+    ///
+    /// This function is only available if the Plugin is able to handle messages that are inside
+    /// the specified `MessageBundle`.
+    fn into_untyped<M: MessageBundle>(self) -> BuiltPlugin
```

Comment:
This is a very odd name and should be changed. Maybe something like `build_plugin`